src/herder/HerderSCPDriver.cpp

```diff
@@ -792,44 +780,74 @@ HerderSCPDriver::valueExternalized(uint64_t slotIndex, Value const& value)
         abort();
     }
 
-    // log information from older ledger to increase the chances that
-    // all messages made it
-    if (slotIndex > 2)
+    // externalize may trigger on older slots:
+    //  * when the current instance starts up
+    //  * when getting back in sync (a gap potentially opened)
+    // in both cases do limited processing on older slots; more importantly,
+    // deliver externalize events to LedgerManager
+    bool newSlot = true;
+    auto currentSlot = mApp.getHerder().getCurrentLedgerSeq();
+    // Allow older messages when in sync
+    if (slotIndex < currentSlot && trackingSCP())
     {
-        logQuorumInformation(slotIndex - 2);
+        newSlot = false;
     }
-
-    if (mCurrentValue)
+    else if (slotIndex <= currentSlot)
     {
-        // stop nomination
-        // this may or may not be the ledger that is currently externalizing
-        // in both cases, we want to stop nomination as:
-        // either we're closing the current ledger (typical case)
-        // or we're going to trigger catchup from history
-        mSCP.stopNomination(mLedgerSeqNominating);
-        mCurrentValue.reset();
+        // Herder already processed slot
+        return;
     }
 
-    if (!mTrackingSCP)
+    // Only update tracking state when newer slot comes in
+    if (newSlot)
     {
-        stateChanged();
-    }
+        // log information from older ledger to increase the chances that
+        // all messages made it
+        if (slotIndex > 2)
+        {
+            logQuorumInformation(slotIndex - 2);
+        }
 
-    mTrackingSCP = std::make_unique<ConsensusData>(slotIndex, b);
+        if (mCurrentValue)
+        {
+            // stop nomination
+            // this may or may not be the ledger that is currently externalizing
+            // in both cases, we want to stop nomination as:
+            // either we're closing the current ledger (typical case)
+            // or we're going to trigger catchup from history
+            mSCP.stopNomination(mLedgerSeqNominating);
+            mCurrentValue.reset();
+        }
 
-    if (!mLastTrackingSCP)
-    {
-        mLastTrackingSCP = std::make_unique<ConsensusData>(*mTrackingSCP);
+        if (!mTrackingSCP)
+        {
+            stateChanged();
+        }
+
+        mTrackingSCP = std::make_unique<ConsensusData>(slotIndex, b);
+
+        if (!mLastTrackingSCP)
+        {
+            mLastTrackingSCP = std::make_unique<ConsensusData>(*mTrackingSCP);
+        }
     }
 
+    auto recordExternalize = [&](bool forceSelf) {
```

Comment:
simpler would have been to just put the old code in the `if (newSlot)` block and add an else block that just calls `mHerder.valueExternalized(slotIndex, b);`

Reply:
I'm not sure how this is simpler: what you are proposing involves calling `valueExternalized` in two places; so now if anything about the function signature or its arguments changes, we'd need to modify it in two places, which makes it slightly more error-prone.

Reply:
it's simpler: it's trading a copy/paste of 1 line visible within 5 lines of code vs a lambda and a flow that is not super straightforward to understand.